File: engineering/concurrency/02-rwlock/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// MyRWMutex 是简易读写锁。
// TODO: 添加必要的状态字段和条件变量。
type MyRWMutex struct {
	mu sync.Mutex
	// TODO: readerCount, writerWaiting, writerActive 等
}

func (rw *MyRWMutex) RLock() {
	rw.mu.Lock()
	defer rw.mu.Unlock()
	// TODO: 如果有活跃写者或等待写者，条件等待；否则增加 readerCount
}

func (rw *MyRWMutex) RUnlock() {
	rw.mu.Lock()
	defer rw.mu.Unlock()
	// TODO: 减少 readerCount，如果是最后一个读者且有写者等待，唤醒写者
}

func (rw *MyRWMutex) Lock() {
	rw.mu.Lock()
	defer rw.mu.Unlock()
	// TODO: 标记有写者等待；如果有活跃读者或写者，条件等待；否则获取写锁
}

func (rw *MyRWMutex) Unlock() {
	rw.mu.Lock()
	defer rw.mu.Unlock()
	// TODO: 释放写锁，唤醒所有等待的读者或下一个写者
}

func main() {
	var rw MyRWMutex
	var wg sync.WaitGroup
	activeReaders := 0
	maxReaders := 0
	var mu sync.Mutex
	writeDuringRead := false

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			rw.RLock()

			mu.Lock()
			activeReaders++
			if activeReaders > maxReaders {
				maxReaders = activeReaders
			}
			mu.Unlock()

			time.Sleep(20 * time.Millisecond)

			mu.Lock()
			activeReaders--
			mu.Unlock()

			rw.RUnlock()
		}(i)
	}

	for i := 0; i < 2; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			rw.Lock()

			mu.Lock()
			if activeReaders > 0 {
				writeDuringRead = true
				fmt.Printf("FAIL: writer %d sees activeReaders=%d\n", id, activeReaders)
			}
			mu.Unlock()

			time.Sleep(20 * time.Millisecond)
			rw.Unlock()
		}(i)
	}

	wg.Wait()
	if !writeDuringRead {
		fmt.Printf("PASS: max concurrent readers = %d\n", maxReaders)
	}
}

File: engineering/concurrency/02-rwlock/skeleton/python/main.py
# 并发模型：threading
import threading
import time

class MyRWMutex:
    def __init__(self):
        self._mu = threading.Lock()
        # TODO: 添加 reader_count, writer_waiting, writer_active 等状态
        # TODO: 添加条件变量

    def rlock(self):
        with self._mu:
            # TODO: 如果有活跃写者，条件等待；否则增加 reader_count
            pass

    def runlock(self):
        with self._mu:
            # TODO: 减少 reader_count，如果是最后一个读者且有写者等待，唤醒写者
            pass

    def lock(self):
        with self._mu:
            # TODO: 标记有写者等待；如果有活跃读者或写者，条件等待；否则获取写锁
            pass

    def unlock(self):
        with self._mu:
            # TODO: 释放写锁，唤醒等待的读者或写者
            pass

def reader(rw, active, max_r, mu, rid):
    rw.rlock()
    with mu:
        active[0] += 1
        if active[0] > max_r[0]:
            max_r[0] = active[0]
    time.sleep(0.02)
    with mu:
        active[0] -= 1
    rw.runlock()

def writer(rw, active, mu, wid):
    rw.lock()
    with mu:
        if active[0] != 0:
            print(f"FAIL: writer {wid} sees activeReaders={active[0]}")
    time.sleep(0.02)
    rw.unlock()

def main():
    rw = MyRWMutex()
    active = [0]
    max_r = [0]
    mu = threading.Lock()
    threads = []

    for i in range(10):
        t = threading.Thread(target=reader, args=(rw, active, max_r, mu, i))
        threads.append(t)
        t.start()

    for i in range(2):
        t = threading.Thread(target=writer, args=(rw, active, mu, i))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    print(f"PASS: max concurrent readers = {max_r[0]}")

if __name__ == "__main__":
    main()
